src/sql/workbench/contrib/query/browser/fileQueryResultsWriter.ts

```diff
@@ -0,0 +1,336 @@
+	}
+
+	public onResultSet(resultSet: ResultSetSummary | ResultSetSummary[]) {
+		this.numQueriesToFormat++;
+		let resultsToAdd: ResultSetSummary[];
+		if (!Array.isArray(resultSet)) {
+			resultsToAdd = [resultSet];
+		} else {
+			resultsToAdd = resultSet.splice(0);
+		}
+
+		if (this.configurationService.getValue<IQueryEditorConfiguration>('queryEditor').results.streaming) {
+			this.addResultSet(resultsToAdd);
+		} else {
+			resultsToAdd = resultsToAdd.filter(e => e.complete);
+			if (resultsToAdd.length > 0) {
+				this.addResultSet(resultsToAdd);
+			}
+		}
+	}
+
+	public async updateResultSet(resultSet: ResultSetSummary | ResultSetSummary[]) {
+		let resultSetSummaries: ResultSetSummary[];
+		if (!Array.isArray(resultSet)) {
+			resultSetSummaries = [resultSet];
+		} else {
+			resultSetSummaries = resultSet.splice(0);
+		}
+
+		if (this.configurationService.getValue<IQueryEditorConfiguration>('queryEditor').results.streaming) {
+			for (let resultSetSummary of resultSetSummaries) {
+				let table = this.tables.find(t => t.resultSet.batchId === resultSetSummary.batchId && t.resultSet.id === resultSetSummary.id);
+				if (table) {
+					await table.updateResult(resultSetSummary);
+				} else {
+					this.logService.warn('Got result set update request for non-existant table');
+				}
+			}
+		} else {
+			resultSetSummaries = resultSetSummaries.filter(e => e.complete);
+			if (resultSetSummaries.length > 0) {
+				this.addResultSet(resultSetSummaries);
+			}
+		}
+	}
+
+	public async onMessage(incomingMessage: IQueryMessage | IQueryMessage[]) {
+		if (isArray(incomingMessage)) {
+			incomingMessage.forEach(m => {
+				if (m.message.includes('Started executing query')) {
+					return;
+				}
+
+				if (m.isError) {
+					this.queryContainsError = true;
+				}
+
+				if (m.message.includes('Total execution time')) {
+					this.closingMessageIncluded = true;
+				}
+
+				this.messages.push(m);
+			});
+		}
+		else {
+			this.messages.push(incomingMessage);
+		}
+
+		if (!this.hasCreatedResultsFile && this.closingMessageIncluded && this.queryContainsError && this.numQueriesToFormat === 0) {
+			await this.createResultsFile();
+		}
+	}
+
+	public reset() {
+		this.messages = [];
+		this.formattedQueryResults = [];
+		this.tables = [];
+		this.numQueriesToFormat = 0;
+		this.queryContainsError = false;
+		this.closingMessageIncluded = false;
+		this.hasCreatedResultsFile = false;
+	}
+
+	private addResultSet(resultSetSummaries: ResultSetSummary[]) {
+		const tables: Array<Table<any>> = [];
+
+		for (const resultSetSummary of resultSetSummaries) {
+			// ensure we aren't adding a resultSet that has already been added
+			if (this.tables.find(t => t.resultSet.batchId === resultSetSummary.batchId && t.resultSet.id === resultSetSummary.id)) {
+				continue;
+			}
+
+			const table = this.instantiationService.createInstance(Table, this.runner, resultSetSummary, this);
+			tables.push(table);
+		}
+
+		this.tables = this.tables.concat(tables);
```

Comment:
using `push` directly on this.tables may be more memory efficient assuming the `Table` objects are relatively large.  If they're small then the memory churn probably doesn't matter much.